File: sdk/ts/src/paths.ts
import { ClientCore } from "./core.js";
import { ensureSuccess } from "./errors.js";
import type { Node, NodeListResponse } from "./types.js";

export interface GetPathOptions {
  pageSize?: number;
  cursor?: string;
  computeRecursiveSizes?: boolean;
}

/**
 * Conflict-handling strategy for file writes. The default is `"error"`:
 * a 409 is returned when a file at the target path already exists.
 *
 * - `"error"` (default) — fail with 409 if the target exists.
 * - `"overwrite"` — replace the existing file in place; node id preserved.
 * - `"rename"` — pick a unique sibling name (foo.jpg → foo-01.jpg) and
 *   create a new file. The returned `node.name`/`node.path` reflect the
 *   actually-used name.
 */
export type FileConflictMode = "error" | "overwrite" | "rename";

export interface PutPathOptions {
  contentType?: string;
  onConflict?: FileConflictMode;
}

export interface PathPutResponse {
  node: Node;
  nodeId: string;
  createdId: string;
  statusCode: number;
}

function getQuery(opts?: GetPathOptions): Record<string, string> {
  const q: Record<string, string> = {};
  if (opts?.pageSize) q["pageSize"] = String(opts.pageSize);
  if (opts?.cursor) q["cursor"] = opts.cursor;
  if (opts?.computeRecursiveSizes) q["computeRecursiveSizes"] = "true";
  return q;
}

function encodeVirtualPath(path: string): string {
  return path
    .split("/")
    .filter((s) => s.length > 0)
    .map(encodeURIComponent)
    .join("/");
}

export class PathsClient {
  constructor(private readonly core: ClientCore) {}

  /**
   * Get a node by virtual path, or list roots when called without a path.
   *
   * - `get()` or `get("")` → lists root mounts (GET /v1/paths/)
   * - `get("data/files")` → metadata for that path (GET /v1/paths/data/files)
   */
  async get(virtualPath?: string, opts?: GetPathOptions): Promise<Node | NodeListResponse> {
    const trimmed = virtualPath?.trim().replace(/^\/+|\/+$/g, "") ?? "";
    if (trimmed === "") {
      return this.core.doJSON<NodeListResponse>("GET", "/v1/paths/", getQuery(opts));
    }
    const encoded = encodeVirtualPath(trimmed);
    return this.core.doJSON<Node>("GET", `/v1/paths/${encoded}`, getQuery(opts));
  }

  /** One-shot upload to a virtual path. */
  async put(virtualPath: string, data: BodyInit, opts?: PutPathOptions): Promise<PathPutResponse> {
    const encoded = encodeVirtualPath(virtualPath.trim());
    if (!encoded) throw new Error("path is required");
    const contentType = opts?.contentType ?? "application/octet-stream";
    const endpoint = `/v1/paths/${encoded}`;
    const resp = await this.core.doRaw("PUT", endpoint, putQuery(opts), data, contentType);
    await ensureSuccess(resp, "PUT", endpoint);
    const node = (await resp.json()) as Node;
    return {
      node,
      nodeId: resp.headers.get("X-Node-Id") ?? "",
      createdId: resp.headers.get("X-Created-Id") ?? "",
      statusCode: resp.status,
    };
  }

  /** One-shot upload returning the raw Response (for relay/proxy). */
  async putRaw(virtualPath: string, data: BodyInit, opts?: PutPathOptions): Promise<Response> {
    const encoded = encodeVirtualPath(virtualPath.trim());
    if (!encoded) throw new Error("path is required");
    const contentType = opts?.contentType ?? "application/octet-stream";
    return this.core.doRaw("PUT", `/v1/paths/${encoded}`, putQuery(opts), data, contentType);
  }
}

function putQuery(opts?: PutPathOptions): Record<string, string> | undefined {
  if (!opts?.onConflict) return undefined;
  return { onConflict: opts.onConflict };
}

File: sdk/filegate/paths.go
package filegate

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

// PathsClient contains path-oriented operations.
type PathsClient struct {
	core *clientCore
}

// GetNodeOptions controls optional directory listing behavior on metadata reads.
type GetNodeOptions struct {
	PageSize              int
	Cursor                string
	ComputeRecursiveSizes bool
}

// FileConflictMode mirrors the server's onConflict vocabulary for file-write
// endpoints (PUT /v1/paths, chunked upload start, transfers). The empty
// string is treated as the server default ("error").
type FileConflictMode string

const (
	ConflictError     FileConflictMode = "error"
	ConflictOverwrite FileConflictMode = "overwrite"
	ConflictRename    FileConflictMode = "rename"
)

// MkdirConflictMode mirrors the server's onConflict vocabulary for mkdir.
// The empty string is treated as the server default ("error").
type MkdirConflictMode string

const (
	MkdirConflictError  MkdirConflictMode = "error"
	MkdirConflictSkip   MkdirConflictMode = "skip"
	MkdirConflictRename MkdirConflictMode = "rename"
)

// PutPathOptions controls one-shot path uploads.
type PutPathOptions struct {
	ContentType string
	// OnConflict selects the server-side behavior on a name collision.
	// Empty defaults to the server default ("error").
	OnConflict FileConflictMode
}

// PathPutResponse is returned by one-shot PUT /v1/paths/* uploads.
type PathPutResponse struct {
	Node       Node
	NodeID     string
	CreatedID  string
	StatusCode int
}

func (o GetNodeOptions) toQuery() url.Values {
	query := url.Values{}
	if o.PageSize > 0 {
		query.Set("pageSize", fmt.Sprintf("%d", o.PageSize))
	}
	if o.Cursor != "" {
		query.Set("cursor", o.Cursor)
	}
	if o.ComputeRecursiveSizes {
		query.Set("computeRecursiveSizes", "true")
	}
	return query
}

// List returns all configured roots as nodes.
func (c PathsClient) List(ctx context.Context) (*NodeListResponse, error) {
	var out NodeListResponse
	if err := c.core.doJSON(ctx, http.MethodGet, "/v1/paths/", nil, nil, "", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get fetches metadata for a virtual path. For directories, children can be paged in the same response.
func (c PathsClient) Get(ctx context.Context, virtualPath string, opts GetNodeOptions) (*Node, error) {
	encodedPath, err := encodeVirtualPath(virtualPath)
	if err != nil {
		return nil, err
	}
	var out Node
	if err := c.core.doJSON(ctx, http.MethodGet, "/v1/paths/"+encodedPath, opts.toQuery(), nil, "", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Put uploads data in one shot to a virtual path. Returns an *APIError on
// non-2xx responses; for passthrough use PutRaw.
func (c PathsClient) Put(ctx context.Context, virtualPath string, data io.Reader, opts PutPathOptions) (*PathPutResponse, error) {
	resp, err := c.PutRaw(ctx, virtualPath, data, opts)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if err := ensureSuccess(resp); err != nil {
		return nil, err
	}

	var node Node
	if err := json.NewDecoder(resp.Body).Decode(&node); err != nil {
		return nil, fmt.Errorf("decode put path response: %w", err)
	}
	return &PathPutResponse{
		Node:       node,
		NodeID:     resp.Header.Get("X-Node-Id"),
		CreatedID:  resp.Header.Get("X-Created-Id"),
		StatusCode: resp.StatusCode,
	}, nil
}

// PutRaw uploads data and returns the raw *http.Response unchanged — including
// non-2xx responses. This is the primitive for relay/passthrough handlers
// that need to forward the upstream status, headers, and body to a
// downstream client without rewriting them. Use Put when you want the
// standard "throw on non-2xx" behavior.
func (c PathsClient) PutRaw(ctx context.Context, virtualPath string, data io.Reader, opts PutPathOptions) (*http.Response, error) {
	encodedPath, err := encodeVirtualPath(virtualPath)
	if err != nil {
		return nil, err
	}
	contentType := opts.ContentType
	if contentType == "" {
		contentType = "application/octet-stream"
	}
	query := url.Values{}
	if opts.OnConflict != "" {
		query.Set("onConflict", string(opts.OnConflict))
	}
	return c.core.doRaw(ctx, http.MethodPut, "/v1/paths/"+encodedPath, query, data, contentType)
}
